### PR TITLE
Finished LinearAlgebra functions, added new features and writed unittests 

### PR DESCRIPTION
**Pull request by 1m8o5s**

### CODE REVIEW

File: GeometryCore/include/LinearAlgebra.h [hunk 1]
```diff
@@ -31,37 +32,46 @@ namespace geom_utils
...
+        return p.x * p2.x + p.y * p2.y + p.z * p2.z;
     }
 
     template<typename T>
     constexpr typename Point3D<T>::coord cross(const Point3D<T>& p, const Point3D<T>& p2)
     {
-        static_assert(false, "cross() is not implemented yet for Point3D");
-        return {};
+        return p.y * p2.z - p.z * p2.y + p.z * p2.x - p.x * p2.z + p.x * p2.y - p.y * p2.x;
     }
```
Comment (VladGribinchuk) [inline GeometryCore/include/LinearAlgebra.h:42]: Cross product for two 3D vectors is another **3D vector**, which is ortogonal to input vectors. Result is not scalar!

File: GeometryCore/src/LinearAlgebra.cpp [hunk 2]
```diff
@@ -4,29 +4,114 @@
...
+        return min_dis;
     }
 
     // Return true if poly and line are colliding with each other.
     // Note: consider collision even if one of line end lies exactly on the polygon contour
-    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line)
-    {
-        return {};
+    bool polygonAndLineSegmentCollides(const Polygon& poly, const LineSegment2D& line) {
+        auto convex = poly.convexHull();
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp]: Why do we calculate convex hull for poly? What will be the result of this function in such case?
image
In fact, line doesn't collide with poly, but it collides with convex hull of poly. Will function return true for that case? (whereas it should return false)

File: GeometryCore/src/LinearAlgebra.cpp [hunk 3]
```diff
@@ -4,29 +4,114 @@
...
+        std::vector<bool> line_collide_with_polygon_side;
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                line_collide_with_polygon_side.push_back(lineSegmentsCollide(line, lie));
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp]: It seems like storing results in vector<bool> is overhead. If we found that some lineSegmentsCollide() return true, we can immediately return true, don't we?

File: GeometryCore/src/LinearAlgebra.cpp [hunk 4]
```diff
@@ -4,29 +4,114 @@
...
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        auto convex = poly.convexHull();
+        std::vector<FPoint2D::coord> distance_from_line_to_polygon_side;
+        #pragma omp parallel for
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                #pragma omp atomic
+                distance_from_line_to_polygon_side.push_back(distanceBetweenLineSegments(line, lie));
+        }
+        FPoint2D::coord min_distance = distance_from_line_to_polygon_side[0];
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp]: You can use std::min_element() for searching minimal element in some range.

File: GeometryCore/src/LinearAlgebra.cpp [hunk 5]
```diff
@@ -4,29 +4,114 @@
...
+    }
+
+    //Return true if polygon is close to line (closer than given gap)
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        auto convex = poly.convexHull();
+        std::vector<FPoint2D::coord> distance_from_line_to_polygon_side;
+        #pragma omp parallel for
+        for (int i = 0; i < convex.size(); ++i) {
+                LineSegment2D lie(convex[i], convex[(i + 1) % convex.size()]);
+                #pragma omp atomic
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp]: Ok. Or you can resize vector to hold enough amount of elements before parallel loop (size should be equal poly.size()) to assign distanceBetweenLineSegments() value for each element in a vector. Thus, you can remove `#pragma omp atomic` to eliminate additional overhead due to synchronization and let simultaneous access to vector.

File: GeometryCore/src/LinearAlgebra.cpp [hunk 6]
```diff
@@ -4,29 +4,114 @@
...
+        }
+        if (std::any_of(line_collide_with_polygon_side.begin(), line_collide_with_polygon_side.end(), [](bool k) {return k; })) {
+            return true;
+        }
+        return false;
+    }
+
+    //Return true if polygon is close to line (closer than given gap)
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        auto convex = poly.convexHull();
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp]: Ditto. Why should we work with convex hull of polygon?

File: GeometryCore/src/LinearAlgebra.cpp [hunk 7]
```diff
@@ -4,29 +4,114 @@
...
+        }
+        return min_distance <= gap;
     }
 
     // Return true if poly and poly2 are close to each other (closer than the provided gap).
     // Note: test vertices of one polygon to see if they are closer than the gap to any lines in other poly.
-    bool polygonsAreAdjacent(const Polygon& poly, const Polygon& poly2, const FPoint2D::coord gap)
-    {
-        return {};
+    bool polygonsAreAdjacent(const Polygon& poly, const Polygon& poly2, const FPoint2D::coord gap) {
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp:103]: Please see my comments to previous polygonAndLineSegmentAreAdjacent() function. My questions/suggestions are the same

**Review by VladGribinchuk — CHANGES_REQUESTED** [review]
In general, good. Test cases are good! But please see some of my concernings/suggestions and fix those.

File: GeometryCore/include/LinearAlgebra.h (hunk 1 above)
Comment (1m8o5s) [reply GeometryCore/include/LinearAlgebra.h:42]: Ok, i will implement this

File: GeometryCore/src/LinearAlgebra.cpp (hunk 2 above)
Comment (1m8o5s) [reply GeometryCore/src/LinearAlgebra.cpp]: Oh, I dont think about this before, I will implement this

File: GeometryCore/src/LinearAlgebra.cpp (hunk 3 above)
Comment (1m8o5s) [reply GeometryCore/src/LinearAlgebra.cpp]: Yes, it will be better

File: GeometryCore/src/LinearAlgebra.cpp (hunk 4 above)
Comment (1m8o5s) [reply GeometryCore/src/LinearAlgebra.cpp]: I will implement this

File: GeometryCore/src/LinearAlgebra.cpp (hunk 5 above)
Comment (1m8o5s) [reply GeometryCore/src/LinearAlgebra.cpp]: Yes, this variant will be better, I will implement this

File: GeometryCore/src/LinearAlgebra.cpp (hunk 6 above)
Comment (1m8o5s) [reply GeometryCore/src/LinearAlgebra.cpp]: Oh, when I do this, I think that polygon contain points, which don`t connected with some others, now i understood that it was wrong think

File: GeometryCore/src/LinearAlgebra.cpp [hunk 8]
```diff
@@ -4,29 +4,107 @@
...
+    bool polygonAndLineSegmentAreAdjacent(const Polygon& poly, const LineSegment2D& line, const FPoint2D::coord gap) {
+        int size = poly.size();
+        std::vector<FPoint2D::coord> distance_from_line_to_polygon_side(size);
+
+        #pragma omp parallel for
+        for (int i = 0; i < size; ++i) {
+                LineSegment2D side(poly[i], poly[(i + 1) % size]);
+                distance_from_line_to_polygon_side[i] = distanceBetweenLineSegments(line, side);
+        }
+        return *std::min_element(distance_from_line_to_polygon_side.begin(), distance_from_line_to_polygon_side.end()) <= gap;
```
Comment (VladGribinchuk) [inline GeometryCore/src/LinearAlgebra.cpp:98]: If poly is empty polygon, then here we will have undefined behavior (dereferencing of iterator not pointing to containers elements)

**Review by VladGribinchuk — CHANGES_REQUESTED** [review]

**Review by VladGribinchuk — COMMENTED** [review]
Good

**Comment by VladGribinchuk** [comment]
If no one complains, I will merge this into main branch

**Review by VladGribinchuk — APPROVED** [review]

**Pull request merged**